### PR TITLE
with_lengthscale

### LINKED ISSUE #326: lengthscale transform
There is currently ScaleTransform, ARDTransform, and LinearTransform. These could all be unified under a LinearTransform with potentially specialised matrices (e.g. a scaled identity matrix or a diagonal matrix for ScaleTransform or ARDTransform, respectively).

Moreover, other kernel/GP packages generally parametrise the kernels with the rather more interpretable _lengthscale_ rather than the inverse lengthscale as KernelFunctions.jl does at the moment: it would be good to have an easy way for users to directly specify a lengthscale (ARD or not).

### PR DESCRIPTION
**Pull request by st--**
To allow for a more natural construction of kernels. Provides the solution discussed in https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/217#issuecomment-760995803 and resolves #326 (the other half of #326 is a duplicate of #107...)

### CODE REVIEW

File: src/transform/with_lengthscale.jl [hunk 1]
```diff
@@ -0,0 +1,34 @@
+"""
+    with_lengthscale(kernel::Kernel, lengthscale::Real)
+    with_lengthscale(kernel::Kernel, lengthscales::AbstractVector{<:Real})
+
+Constructs a transformed kernel with `lengthscale`. If `lengthscales` is passed
+as a vector instead, constructs an "ARD" kernel with different lengthscales for
+each dimension.
```
Comment (willtebbutt) [inline src/transform/with_lengthscale.jl:7]:
```suggestion
Constructs a transformed kernel with `lengthscale`.
If a vector `lengthscales` is passed instead, constructs an "ARD" kernel with different lengthscales for each dimension.
```
Minor grammar thing -- I think this way makes more sense to me.

File: src/transform/ardtransform.jl [hunk 2]
```diff
@@ -15,6 +15,7 @@ true
 struct ARDTransform{Tv<:AbstractVector{<:Real}} <: Transform
     v::Tv
 end
+Base .== (a::ARDTransform, b::ARDTransform) = a.v == b.v
```
Comment (devmotion) [inline src/transform/ardtransform.jl:18]:
```suggestion
Base.:(==)(a::ARDTransform, b::ARDTransform) = a.v == b.v
```

File: src/transform/scaletransform.jl [hunk 3]
```diff
@@ -15,6 +15,7 @@ true
 struct ScaleTransform{T<:Real} <: Transform
     s::Vector{T}
 end
+Base .== (a::ScaleTransform, b::ScaleTransform) = a.s == b.s
```
Comment (devmotion) [inline src/transform/scaletransform.jl:18]:
```suggestion
Base.:(==)(a::ScaleTransform, b::ScaleTransform) = a.s == b.s
```

File: src/transform/with_lengthscale.jl [hunk 4]
```diff
@@ -0,0 +1,33 @@
+"""
+    with_lengthscale(kernel::Kernel, lengthscale::Real)
+    with_lengthscale(kernel::Kernel, lengthscales::AbstractVector{<:Real})
```
Comment (devmotion) [inline src/transform/with_lengthscale.jl:3]: It might be a bit easier, both to read and write, if one adds separate docstrings to the two methods. They will both show up if one looks up `with_lengthscale`.

File: src/transform/with_lengthscale.jl [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+Constructs a transformed kernel with `lengthscale`.
+If a vector `lengthscales` is passed instead, constructs an "ARD" kernel with different lengthscales for each dimension.
```
Comment (devmotion) [inline src/transform/with_lengthscale.jl:6]: The convention is to use the imperative form in the first line of the docstring:
```suggestion
Construct a transformed kernel with `lengthscale`.
```

File: src/transform/with_lengthscale.jl [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+julia> SqExponentialKernel() ∘ ARDTransform(inv.(ℓ)) == with_lengthscale(SqExponentialKernel(), ℓ)
+true
+```
+
+"""
+function with_lengthscale(kernel::Kernel, lengthscale::Real)
+    return compose(kernel, ScaleTransform(inv(lengthscale)))
+end
+function with_lengthscale(kernel::Kernel, lengthscales::AbstractVector{<:Real})
+    return compose(kernel, ARDTransform(inv.(lengthscales)))
```
Comment (devmotion) [inline src/transform/with_lengthscale.jl:32]: The broadcasting machinery is not needed it seems:
```suggestion
    return kernel ∘ ARDTransform(map(inv, lengthscales))
```

Comment (st--) [reply src/transform/with_lengthscale.jl:32]: why would you prefer `map` instead of the broadcasting?

Comment (devmotion) [reply src/transform/with_lengthscale.jl:32]: The broadcasting machinery is very heavy and puts a lot of additional stress on the compiler, so in general I avoid it if there is no actual broadcasting of dimensions needed.

File: src/kernels/transformedkernel.jl [hunk 7]
```diff
@@ -72,6 +72,10 @@ Base.:∘(k::TransformedKernel, t::Transform) = TransformedKernel(k.kernel, k.tr
 Base.:∘(k::Kernel, ::IdentityTransform) = k
 Base.:∘(k::TransformedKernel, ::IdentityTransform) = k
 
+function Base.isequal(k::TransformedKernel, k2::TransformedKernel)
+    return isequal(k.kernel, k2.kernel) && isequal(k.transform, k2.transform)
+end
```
Comment (devmotion) [inline src/kernels/transformedkernel.jl:77]: If one defines `isequal` one has to define `hash` as well since otherwise the two are not consistent. In general, if one does not care about floating point comparisons one should just define `==` instead of `isequal` since `isequal` falls back to `==`. Here it probably would make sense to define both since it would lead to different comparisons of the kernel parameters and parameters of the transform.

Basically, most often one wants the recursive implementation in https://github.com/andrewcooke/AutoHashEquals.jl/issues/18.

Comment (st--) [reply src/kernels/transformedkernel.jl:77]: I don't follow why one cannot define `isequal` without defining `hash`; it can be easy to check containers for equality without being able to construct a hash function. Based on the docstrings of `isequal` and `==`, it seems like the former is the one to use for containers. But in any case that seems out of scope of this PR.

Comment (devmotion) [reply src/kernels/transformedkernel.jl:77]: > I don't follow why one cannot define isequal without defining hash; it can be easy to check containers for equality without being able to construct a hash function.

It would lead to an inconsistency. It is explained in the docstring of `isequal` (and a bit shorter in the one for `==`):

```julia
  isequal is the comparison function used by hash tables (Dict). isequal(x,y) must imply that hash(x) == hash(y).

  This typically means that types for which a custom == or isequal method exists must implement a corresponding hash method (and vice versa). Collections typically implement
  isequal by calling isequal recursively on all contents.
```

> Based on the docstrings of isequal and ==, it seems like the former is the one to use for containers.

This is not correct, e.g., both are defined for arrays: https://github.com/JuliaLang/julia/blob/d85d8a91b79021668a855ab496ff3d79c2de5f4b/base/abstractarray.jl#L1842 and https://github.com/JuliaLang/julia/blob/d85d8a91b79021668a855ab496ff3d79c2de5f4b/base/abstractarray.jl#L1818 The difference is the floating point comparisons, and hence for anything involving floating point numbers it makes sense to define both.

> But in any case that seems out of scope of this PR.

Just defining `isequal` leads to incorrect behaviour for dictionaries, so at least one should add `hash` as well in this PR. And then I guess one can just define `==` as well?

Comment (st--) [reply src/kernels/transformedkernel.jl:77]: > Just defining isequal leads to incorrect behaviour for dictionaries,

could you give an example of what you mean ?

> so at least one should add hash as well in this PR. And then I guess one can just define == as well?

There were already plenty of `isequal` implementations in the codebase; it was just the one for `TransformedKernel` that was missing to be able to write the doctests. I think "add `hash` for all of them" is beyond the scope of this PR. If you don't want this PR merged without adding all of those, please feel free to take over and do the work.

Comment (devmotion) [reply src/kernels/transformedkernel.jl:77]: > could you give an example of what you mean ?

Dictionaries use the `hash` function to search for keys but not `isequal`. Thus if the `hash` function is not consistent with `isequal`, objects that are equal according to `isequal` can be considered as different by the `hash` function, and hence the same key could appear multiple times. And vice versa, objects might seem to be equal according to `hash` and their values are replaced even though actually they are different according to `isequal`.

> There were already plenty of isequal implementations in the codebase; it was just the one for TransformedKernel that was missing to be able to write the doctests. I think "add hash for all of them" is beyond the scope of this PR. If you don't want this PR merged without adding all of those, please feel free to take over and do the work.

Existing implementations should definitely not be fixed in this PR. However, I think we should not add a broken/inconsistent implementation of equality checks in this PR. In my opinion, either we should not add any equality checks at all and use different tests (e.g., just check the types or the lengthscale explicitly) or we should define the equality checks correctly (i.e., add `isequal`, `hash`, and `==`).

Comment (st--) [reply src/kernels/transformedkernel.jl:77]: please take over the PR then, this was how far I was prepared to take it:)

Comment (devmotion) [reply src/kernels/transformedkernel.jl:77]: I opened https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/336, I guess it is cleaner to just not define `isequal` in this PR.

**Comment by devmotion** [comment]
Fixed by https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/pull/336.
